crates/subspace-farmer/src/single_plot_farm.rs

```diff
@@ -733,23 +733,44 @@ impl<RC: RpcClient> VerifyingPlotter<RC> {
             return Err(PiecesVerificationError::InvalidFarmerProtocolInfo);
         }
 
-        for (piece, piece_index) in pieces.as_pieces().zip(piece_indexes) {
-            let segment_index: u64 = piece_index / records_per_segment as u64;
-            let position: u64 = piece_index % records_per_segment as u64;
+        let segment_indexes = piece_indexes
+            .iter()
+            .map(|piece_index| piece_index / records_per_segment as u64)
+            .collect::<Vec<_>>();
+
+        let roots = self
+            .verification_client
+            .records_roots(segment_indexes)
+            .await
+            .map_err(|err| PiecesVerificationError::RpcError(err))?;
+
+        let verified_roots = roots
+            .into_iter()
+            .zip(piece_indexes.iter())
+            .map(|(root, piece_index)| {
+                if let Some(root) = root {
+                    Ok(root)
+                } else {
+                    error!(?piece_index, "No records root found for piece_index.");
 
-            let root = self
-                .verification_client
-                .records_root(segment_index)
-                .await
-                .map_err(|err| PiecesVerificationError::RpcError(err))?
-                .ok_or(PiecesVerificationError::NoRecordsRootFound)?;
+                    Err(PiecesVerificationError::NoRecordsRootFound)
+                }
+            })
+            .collect::<Result<Vec<Sha256Hash>, PiecesVerificationError>>()?;
+
+        for ((piece, piece_index), root) in
+            pieces.as_pieces().zip(piece_indexes).zip(verified_roots)
+        {
+            let position: u64 = piece_index % records_per_segment as u64;
 
             if !is_piece_valid(
                 piece,
                 root,
                 position as usize,
                 self.farmer_protocol_info.record_size as usize,
             ) {
+                error!(?piece_index, "Piece validation failed.");
```

Comment:
I believe you can remove `?` here. The thing is `tracing` has typed values and there are untyped ones like `Debug` and `Display` (also for `std::error::Error` implementors I believe)